Lovely Rita: EDA

In [1]:
import math
from collections import defaultdict
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import itertools
import random
from dateutil import relativedelta

pd.options.display.max_rows = 100
pd.options.display.max_columns = 100

Static Variables

In [16]:
data_files_01 = [
    "../data/2012complete-output.csv", 
    "../data/2013complete-output.csv", 
    "../data/2014complete-output.csv", 
    "../data/2015complete-output.csv"
]

dtype_01 = {
    'street': str, 
    'city': str, 
    'state': str, 
    'ticket_number': str, 
    'ticket_issue_date': str, 
    'ticket_issue_time': str, 
    'violation_external_code': str,
    'violation_desc_long': str, 
    'street_no': str, 
    'street_name': str, 
    'street_suffix': str,
    'fine_amount': str, 
    'badge__': str, 
    'Unnamed: 13': str, 
    '[sequence]': int, 
    '[summary]': str,
    '[addressee]': str, 
    '[delivery_line_1]': str, 
    '[delivery_line_2]': str,
    '[city_name]': str,
    '[state_abbreviation]': str, 
    '[full_zipcode]': str, 
    '[notes]': str,
    '[county_name]': str, 
    '[rdi]': str, 
    '[latitude]': float, 
    '[longitude]': float,
    '[precision]': str, 
    '[dpv_match_code]': str, 
    '[dpv_footnotes]': str, 
    '[footnotes]': str,
    '[county_fips]': str, 
    '[record_type]': str, 
    '[zip_type]': str, 
    '[carrier_route]': str,
    '[congressional_district]': str, 
    '[building_default_indicator]': str, 
    '[elot_sequence]': str, 
    '[elot_sort]': str, 
    '[time_zone]': str, 
    '[utc_offset]': str,
    '[dst]': str, 
    '[dpv_cmra]': str, 
    '[ews_match]': str, 
    '[lacslink_indicator]': str,
    '[lacslink_code]': str, 
    '[suitelink_match]': str, 
    '[dpv_vacant]': str, 
    '[active]': str,
    '[urbanization]': str, 
    '[primary_number]': str, 
    '[street_name]': str,
    '[street_predirection]': str, 
    '[street_postdirection]': str,
    '[street_suffix]': str, 
    '[secondary_number]': str, 
    '[secondary_designator]': str,
    '[extra_secondary_number]': str, 
    '[extra_secondary_designator]': str,
    '[pmb_designator]': str, 
    '[pmb_number]': str, 
    '[default_city_name]': str,
    '[zipcode]': str, 
    '[plus4_code]': str, 
    '[delivery_point]': str,
    '[delivery_point_check_digit]': str, 
    '[last_line]': str,
    '[delivery_point_barcode]': str
}

column_map_01 = {
    'street': 'Street', 
    'city': 'City', 
    'state': 'State', 
    'ticket_number': 'Ticket Number', 
    'ticket_issue_date': 'Ticket Issue Date', 
    'ticket_issue_time': 'Ticket Issue Time', 
    'violation_external_code': 'Violation External Code',
    'violation_desc_long': 'Violation Description Long', 
    'street_no': 'Street Number', 
    'street_name': 'Street Name', 
    'street_suffix': 'Street Suffix',
    'fine_amount': 'Fine Amount', 
    'badge__': 'Badge Number', 
    'Unnamed: 13': 'Unnamed: 13', 
    '[sequence]': 'Sequence', 
    '[summary]': 'Summary',
    '[addressee]': 'Addressee', 
    '[delivery_line_1]': 'Delivery Line 1', 
    '[delivery_line_2]': 'Delivery Line 2',
    '[city_name]': 'City Name',
    '[state_abbreviation]': 'State Abbreviation', 
    '[full_zipcode]': 'Full Zipcode', 
    '[notes]': 'Notes',
    '[county_name]': 'County Name', 
    '[rdi]': 'RCI', 
    '[latitude]': 'Latitude', 
    '[longitude]': 'Longitude',
    '[precision]': 'Precision', 
    '[dpv_match_code]': 'DPV Match Code', 
    '[dpv_footnotes]': 'DPV Footnotes', 
    '[footnotes]': 'Footnotes',
    '[county_fips]': 'County FIPS', 
    '[record_type]': 'Record Type', 
    '[zip_type]': 'Zip Type', 
    '[carrier_route]': 'Carrier Route',
    '[congressional_district]': 'Congressional District', 
    '[building_default_indicator]': 'Building Default Indicator', 
    '[elot_sequence]': 'eLot Sequence', 
    '[elot_sort]': 'eLot Sort', 
    '[time_zone]': 'Time Zone', 
    '[utc_offset]': 'UTC Offset',
    '[dst]': 'DST', 
    '[dpv_cmra]': 'DPV Camera', 
    '[ews_match]': 'EWS Match', 
    '[lacslink_indicator]': 'LACS Link Indicator',
    '[lacslink_code]': 'LACS Link Code', 
    '[suitelink_match]': 'Suite Link Match', 
    '[dpv_vacant]': 'DPV Vacant', 
    '[active]': 'Active',
    '[urbanization]': 'Urbanization', 
    '[primary_number]': 'Primary Number', 
    '[street_name]': 'Street Name 2',
    '[street_predirection]': 'Street Predirection', 
    '[street_postdirection]': 'Street Postdirection',
    '[street_suffix]': 'Street Suffix 2', 
    '[secondary_number]': 'Secondary Number', 
    '[secondary_designator]': 'Secondary Designator',
    '[extra_secondary_number]': 'Extra Secondary Number', 
    '[extra_secondary_designator]': 'Extra Secondary Designator',
    '[pmb_designator]': 'PMB Designator', 
    '[pmb_number]': 'PMB Number', 
    '[default_city_name]': 'Default City Name',
    '[zipcode]': 'Zipcode', 
    '[plus4_code]': 'Plus 4 Code', 
    '[delivery_point]': 'Delivery Point',
    '[delivery_point_check_digit]': 'Delivery Point Check Digit', 
    '[last_line]': 'Last Line',
    '[delivery_point_barcode]': 'Delivery Point Barcode'
}

data_files_02 = [
    "../data/2016complete-output.csv"
]

dtype_02 = {
    'street': str, 
    'city': str, 
    'state': str, 
    'Ticket Number': str, 
    'Ticket Issue Date': str, 
    'Ticket Issue Time': str, 
    'Violation External Code': str,
    'Violation Desc Long': str, 
    'Street No': str, 
    'Street Name': str, 
    'Street Suffix': str,
    'Fine Amount': str, 
    'Badge #': str, 
    'Unnamed: 13': str, 
    '[sequence]': int, 
    '[summary]': str,
    '[addressee]': str, 
    '[delivery_line_1]': str, 
    '[delivery_line_2]': str,
    '[city_name]': str,
    '[state_abbreviation]': str, 
    '[full_zipcode]': str, 
    '[notes]': str,
    '[county_name]': str, 
    '[rdi]': str, 
    '[latitude]': float, 
    '[longitude]': float,
    '[precision]': str, 
    '[dpv_match_code]': str, 
    '[dpv_footnotes]': str, 
    '[footnotes]': str,
    '[county_fips]': str, 
    '[record_type]': str, 
    '[zip_type]': str, 
    '[carrier_route]': str,
    '[congressional_district]': str, 
    '[building_default_indicator]': str, 
    '[elot_sequence]': str, 
    '[elot_sort]': str, 
    '[time_zone]': str, 
    '[utc_offset]': str,
    '[dst]': str, 
    '[dpv_cmra]': str, 
    '[ews_match]': str, 
    '[lacslink_indicator]': str,
    '[lacslink_code]': str, 
    '[suitelink_match]': str, 
    '[dpv_vacant]': str, 
    '[active]': str,
    '[urbanization]': str, 
    '[primary_number]': str, 
    '[street_name]': str,
    '[street_predirection]': str, 
    '[street_postdirection]': str,
    '[street_suffix]': str, 
    '[secondary_number]': str, 
    '[secondary_designator]': str,
    '[extra_secondary_number]': str, 
    '[extra_secondary_designator]': str,
    '[pmb_designator]': str, 
    '[pmb_number]': str, 
    '[default_city_name]': str,
    '[zipcode]': str, 
    '[plus4_code]': str, 
    '[delivery_point]': str,
    '[delivery_point_check_digit]': str, 
    '[last_line]': str,
    '[delivery_point_barcode]': str
}

column_map_02 = {
    'street': 'Street', 
    'city': 'City', 
    'state': 'State', 
    'Ticket Number': 'Ticket Number', 
    'Ticket Issue Date': 'Ticket Issue Date', 
    'Ticket Issue Time': 'Ticket Issue Time', 
    'Violation External Code': 'Violation External Code',
    'Violation Desc Long': 'Violation Description Long', 
    'Street No': 'Street Number', 
    'Street Name': 'Street Name', 
    'Street Suffix': 'Street Suffix',
    'Fine Amount': 'Fine Amount', 
    'Badge #': 'Badge Number', 
    'Unnamed: 13': 'Unnamed: 13', 
    '[sequence]': 'Sequence', 
    '[summary]': 'Summary',
    '[addressee]': 'Addressee', 
    '[delivery_line_1]': 'Delivery Line 1', 
    '[delivery_line_2]': 'Delivery Line 2',
    '[city_name]': 'City Name',
    '[state_abbreviation]': 'State Abbreviation', 
    '[full_zipcode]': 'Full Zipcode', 
    '[notes]': 'Notes',
    '[county_name]': 'County Name', 
    '[rdi]': 'RCI', 
    '[latitude]': 'Latitude', 
    '[longitude]': 'Longitude',
    '[precision]': 'Precision', 
    '[dpv_match_code]': 'DPV Match Code', 
    '[dpv_footnotes]': 'DPV Footnotes', 
    '[footnotes]': 'Footnotes',
    '[county_fips]': 'County FIPS', 
    '[record_type]': 'Record Type', 
    '[zip_type]': 'Zip Type', 
    '[carrier_route]': 'Carrier Route',
    '[congressional_district]': 'Congressional District', 
    '[building_default_indicator]': 'Building Default Indicator', 
    '[elot_sequence]': 'eLot Sequence', 
    '[elot_sort]': 'eLot Sort', 
    '[time_zone]': 'Time Zone', 
    '[utc_offset]': 'UTC Offset',
    '[dst]': 'DST', 
    '[dpv_cmra]': 'DPV Camera', 
    '[ews_match]': 'EWS Match', 
    '[lacslink_indicator]': 'LACS Link Indicator',
    '[lacslink_code]': 'LACS Link Code', 
    '[suitelink_match]': 'Suite Link Match', 
    '[dpv_vacant]': 'DPV Vacant', 
    '[active]': 'Active',
    '[urbanization]': 'Urbanization', 
    '[primary_number]': 'Primary Number', 
    '[street_name]': 'Street Name 2',
    '[street_predirection]': 'Street Predirection', 
    '[street_postdirection]': 'Street Postdirection',
    '[street_suffix]': 'Street Suffix 2', 
    '[secondary_number]': 'Secondary Number', 
    '[secondary_designator]': 'Secondary Designator',
    '[extra_secondary_number]': 'Extra Secondary Number', 
    '[extra_secondary_designator]': 'Extra Secondary Designator',
    '[pmb_designator]': 'PMB Designator', 
    '[pmb_number]': 'PMB Number', 
    '[default_city_name]': 'Default City Name',
    '[zipcode]': 'Zipcode', 
    '[plus4_code]': 'Plus 4 Code', 
    '[delivery_point]': 'Delivery Point',
    '[delivery_point_check_digit]': 'Delivery Point Check Digit', 
    '[last_line]': 'Last Line',
    '[delivery_point_barcode]': 'Delivery Point Barcode'
}

delimiter = ","

address_file_01 = "../data/addresses_01.txt"
address_file_02 = "../data/addresses_02.txt"

columns_to_drop = [
    "Addressee", 
    "Delivery Line 2", 
    "EWS Match", 
    "Extra Secondary Number", 
    "PMB Designator", 
    "PMB Number", 
    "Suite Link Match", 
    "Unnamed: 13", 
    "Urbanization"
]

Functions

In [56]:
def get_data(data_files, dtype, column_map, delimiter):
    df = pd.concat([pd.read_csv(f, dtype=dtype, delimiter=delimiter) for f in data_files]).reset_index(drop=True)
    df.rename(columns=column_map, inplace=True)
    return df

def get_column_report(df):
    column_report = []
    for column in df.columns:
        unique = df[column].unique()
        sample = np.nan
        for value in unique:
            if value is not np.nan:
                sample = value
                break
        nans = df[column].isnull().sum()
        pct_nan = 100. * nans / df.shape[0]
        column_report.append([column, df[column].dtype, len(unique), sample, nans, pct_nan])
    columns=["Column Name", "Data Type", "Unique Count", "Sample Value", "NaNs", "% NaN"]
    column_report = pd.DataFrame(column_report, columns=columns).round(2)
    column_report.sort_values(by="NaNs", inplace=True)
    return column_report

def get_uniques(df):
    for column in df.columns:
        print column, df[column].unique()
        
def output_addresses(condition, file_out):
    addresses = set()
    for i, item in (df[condition]["Street"] + " " + df[condition]["City"] + " " + df[condition]["State"]).iteritems():
        addresses.add(" ".join(item.lower().split()))
    with open(file_out, 'w') as output:
        for address in addresses:
            output.write(address + '\n')
    return addresses

def convert_time_str_to_min(time_str):
    time_elements = time_str.split(":")
    hours = int(time_elements[0])
    minutes = int(time_elements[1])
    return minutes + (hours * 60)

def convert_min_to_time_str(minutes):
    minutes_str = str(minutes % 60)
    if len(minutes_str) == 1:
        minutes_str = '0' + minutes_str
    return str((minutes / 60) % 24) + ":" + minutes_str
    
def get_time_diff(time_start, time_end):
    time_1 = convert_time_str_to_min(time_start)
    time_2 = convert_time_str_to_min(time_end)
    if time_2 < time_1:
        time_2 += 24 * 60
    return time_2 - time_1

Load Raw Data and Merge

In [4]:
df_01 = get_data(data_files_01, dtype_01, column_map_01, delimiter)
print df_01.shape

(1319187, 68)


In [5]:
df_02 = get_data(data_files_02, dtype_02, column_map_02, delimiter)
df_02 = df_02[:-1] # last row is junk
print df_02.shape

(326625, 68)


In [6]:
df = pd.concat([df_01, df_02]).reset_index(drop=True)
print df.shape
df_01 = None
df_02 = None

(1645812, 68)


In [7]:
get_column_report(df)

Column Name Data Type  Unique Count  \
55           Ticket Issue Date    object          1827   
38                         RCI    object             3   
36                   Precision    object             5   
45                      Street    object        174339   
31                   Longitude   float64         10575   
30                    Latitude   float64          9324   
54                     Summary    object             5   
43                       State    object             2   
57               Ticket Number    object       1645812   
58                   Time Zone    object             2   
62  Violation Description Long    object            87   
42                    Sequence     int64        347427   
63     Violation External Code    object            98   
2                 Badge Number    object          1461   
5                         City    object             2   
16             Delivery Line 1    object        164654   
32                       Notes    object            64   
29                   Last Line    object         20025   
11               DPV Footnotes    object             9   
6                    City Name    object             3   
44          State Abbreviation    object             2   
46                 Street Name    object          8579   
56           Ticket Issue Time    object          2872   
24                 Fine Amount    object            68   
48               Street Number    object         10859   
51               Street Suffix    object            18   
7       Congressional District    object             3   
47               Street Name 2    object          1263   
8                  County FIPS    object             2   
4                Carrier Route    object            53   
9                  County Name    object             2   
59                  UTC Offset    object             2   
39                 Record Type    object             4   
14                         DST    object             2   
26                Full Zipcode    object         20024   
19      Delivery Point Barcode    object         83474   
64                    Zip Type    object             3   
65                     Zipcode    object            18   
37              Primary Number    object          9223   
12              DPV Match Code    object             5   
35                 Plus 4 Code    object          4324   
52             Street Suffix 2    object            21   
0                       Active    object             3   
67                   eLot Sort    object             3   
13                  DPV Vacant    object             3   
66               eLot Sequence    object           377   
18              Delivery Point    object           101   
20  Delivery Point Check Digit    object            11   
10                  DPV Camera    object             3   
25                   Footnotes    object            38   
3   Building Default Indicator    object             2   
50         Street Predirection    object             5   
15           Default City Name    object             3   
40        Secondary Designator    object            11   
27              LACS Link Code    object             3   
28         LACS Link Indicator    object             3   
49        Street Postdirection    object             5   
41            Secondary Number    object            50   
22  Extra Secondary Designator    object             2   
1                    Addressee    object             1   
61                Urbanization    object             1   
21                   EWS Match    object             1   
17             Delivery Line 2    object             1   
53            Suite Link Match    object             1   
23      Extra Secondary Number    object             1   
34                  PMB Number    object             1   
60                 Unnamed: 13    object             1   
33              PMB Designator    object             1   

                

Unique IDs:
Ticket Number
Sequence (with Year)

Empty columns:
Addressee
Delivery Line 2
EWS Match
Extra Secondary Number
PMB Designator
PMB Number
Suite Link Match
Unnamed: 13
Urbanization

Invariant information columns:
County FIPS
County Name
State
State Abbreviation
Time Zone
UTC Offset

'> 80% NaN column:
Building Default Indicator
Default City Name
Extra Secondary Designator
LACS Link Code
LACS Link Indicator
Secondary Designator
Secondary Number
Street Postdirection
Street Predirection


Location Parity Analysis

Location columns:
Street
City
State

Street Number
Street Name
Street Suffix

Delivery Line 1
City Name
State Abbreviation
Full Zipcode

Latitude
Longitude

Primary Number
Street Name 2
Street Suffix 2
Zipcode
Plus 4 Code
Last Line


In [8]:
count = 0
for i, row in df.iterrows():
    street_tokens = row["Street"].strip().split()
    s_zeroes = []
    if len(street_tokens) > 0:
        if len(street_tokens[0]) > 0:
            s_zero = street_tokens[0]
            s_zeroes.append(s_zero)
            if s_zero == 'ONE':
                s_zero = '1'
                s_zeroes.append(s_zero)
            elif s_zero == 'TENTH':
                s_zero = '10'
                s_zeroes.append(s_zero)
            elif (s_zero == 'SOUTHWEST') | (s_zero == '#'):
                s_zeroes.append(street_tokens[1])
            elif not s_zero[0].isdigit():
                s_zero = s_zero[1:]
                s_zeroes.append(s_zero)
            if (len(street_tokens) > 3):
                s_one = street_tokens[1]
                if len(s_one) == 2:
                    if not s_one[0].isalpha():
                        s_one = s_one[1:]
                    elif not s_one[1].isalpha():
                        s_one = s_one[:1]
                if (len(s_one) == 1) & s_one.isalpha() & (s_one not in ['N', 'S', 'E', 'W']):
                    s_zeroes.append(s_zero + s_one)
                    s_zeroes.append(s_zero + " " + s_one)
    if type(row["Primary Number"]) != float:
        if row["Primary Number"].strip() not in s_zeroes:
            count += 1
            print row["Primary Number"], " | ", row["Street"], " | ", s_zeroes
print count

3236  |  3238 GRAND AVE  |  ['3238']
503  |  501 FOREST ST  |  ['501']
3236  |  3238 GRAND AVE  |  ['3238']
503  |  501 FOREST ST  |  ['501']
503  |  501 FOREST ST  |  ['501']
503  |  501 FOREST ST  |  ['501']
3236  |  3238 GRAND AVE  |  ['3238']
503  |  501 FOREST ST  |  ['501']
503  |  501 FOREST ST  |  ['501']
3236  |  3238 GRAND AVE  |  ['3238']
503  |  501 FOREST ST  |  ['501']
11


There are 11 rows in which "Primary Number" is not in parity with "Street".

In [9]:
count = 0
for i, row in df.iterrows():
    if type(row["Street Name 2"]) != float:
        street_tokens = row["Street"].strip().split()
        street_names = []
        street_name = " ".join(street_tokens[1:-1]).lower()
        street_names.append(street_name)
        if street_tokens[-1] == 'BROADWAY':
            street_names.append("broadway")
        elif street_tokens[-1] == "INTERNATIONAL":
            street_names.append("international")
        elif street_name == "martin l king jr":
            street_names.append("martin luther king jr")
        elif street_name == "sancarlos":
            street_names.append("san carlos")
        elif street_name == "agua vista":
            street_names.append("aqua vista")
        elif street_name == "iselton":
            street_names.append("isleton")
        elif street_name == "lake shore":
            street_names.append("lakeshore")
        elif street_name == "bay view":
            street_names.append("bayview")
        elif (street_name == "grand ave") | (street_name == "west grand"):
            street_names.append("grand")
        elif street_name == "thomas l":
            street_names.append("thomas l berkley")
        elif len(street_tokens) > 1:
            if street_tokens[1] == "MC":
                street_names.append((street_tokens[1] + street_tokens[2]).lower())
        elif street_name == "benevenue":
            street_names.append("benvenue")
        elif street_name == "23":
            street_names.append("23rd")
        if row["Street Name 2"].lower() not in street_names:
            count += 1
            # print row["Street Name 2"], " | ", row["Street"], " | ", street_names
print count

31528


Street Name Parity Check: Still In Progress

Export unique addresses for Geocode lookup
Two priority orders will be: 1) addresses without Lat/Long data and 2) addresses with Lat/Long data.

In [25]:
addresses_01 = output_addresses((df["Latitude"]==0), address_file_01)
addresses_02 = output_addresses((df["Latitude"]!=0), address_file_02)
print len(set(addresses_01)), len(set(addresses_02))

33835 140490


DateTime Analysis

In [11]:
print df["Ticket Issue Date"].isnull().sum(), df["Ticket Issue Time"].isnull().sum()

0 119


In [12]:
(np.diff(np.where(df["Ticket Issue Time"].isnull()))<=1).sum()

8

There are 0 missing dates and 119 missing times.
Since the data is recorded in temporal order, the times in the neighboring rows represent a window in which the missing time exists.  We can impute the missing time values as the mean time of the neighboring rows.  There are 8 cases where missing values are next to eachother.  For those, we can find the next existing value, and divide the time evenly to fill in the missing values.

In [63]:
for i in np.where(df["Ticket Issue Time"].isnull())[0]:
    prev_value = np.nan
    prev_index = i
    while type(prev_value) != str:
        prev_index -= 1
        prev_value = df.iloc[prev_index]["Ticket Issue Time"]
    #print type(prev_value)
    #print prev_value
    next_value = np.nan
    next_index = i
    while type(next_value) != str:
        next_index += 1
        next_value = df.iloc[next_index]["Ticket Issue Time"]
    diff = get_time_diff(prev_value, next_value)
    split_count = next_index-prev_index
    prev_minutes = convert_time_str_to_min(prev_value)
    for _ in range(split_count):
        prev_minutes += (diff / split_count)
        df.at[i, "Ticket Issue Time"] = convert_min_to_time_str(prev_minutes)

In [65]:
df["Ticket Issue DateTime"] = pd.to_datetime(df["Ticket Issue Date"] + " " + df["Ticket Issue Time"])

In [66]:
get_column_report(df)

Column Name       Data Type  Unique Count  \
68       Ticket Issue DateTime  datetime64[ns]        832073   
45                      Street          object        174339   
38                         RCI          object             3   
36                   Precision          object             5   
31                   Longitude         float64         10575   
30                    Latitude         float64          9324   
54                     Summary          object             5   
55           Ticket Issue Date          object          1827   
42                    Sequence           int64        347427   
57               Ticket Number          object       1645812   
58                   Time Zone          object             2   
62  Violation Description Long          object            87   
63     Violation External Code          object            98   
56           Ticket Issue Time          object          2872   
43                       State          object             2   
2                 Badge Number          object          1461   
5                         City          object             2   
11               DPV Footnotes          object             9   
16             Delivery Line 1          object        164654   
44          State Abbreviation          object             2   
32                       Notes          object            64   
29                   Last Line          object         20025   
6                    City Name          object             3   
46                 Street Name          object          8579   
24                 Fine Amount          object            68   
48               Street Number          object         10859   
51               Street Suffix          object            18   
14                         DST          object             2   
59                  UTC Offset          object             2   
7       Congressional District          object             3   
8                  County FIPS          object             2   
4                Carrier Route          object            53   
9                  County Name          object             2   
47               Street Name 2          object          1263   
39                 Record Type          object             4   
64                    Zip Type          object             3   
26                Full Zipcode          object         20024   
65                     Zipcode          object            18   
19      Delivery Point Barcode          object         83474   
37              Primary Number          object          9223   
35                 Plus 4 Code          object          4324   
12              DPV Match Code          object             5   
52             Street Suffix 2          object            21   
66               eLot Sequence          object           377   
0                       Active          object             3   
67                   eLot Sort          object             3   
10                  DPV Camera          object             3   
20  Delivery Point Check Digit          object            11   
18              Delivery Point          object           101   
13                  DPV Vacant          object             3   
25                   Footnotes          object            38   
3   Building Default Indicator          object             2   
50         Street Predirection          object             5   
15           Default City Name          object             3   
40        Secondary Designator          object            11   
28         LACS Link Indicator          object             3   
27              LACS Link Code          object             3   
49        Street Postdirection          object             5   
41            Secondary Number          object            50   
22  Extra Secondary Designator          object             2   
1                    Addressee          object             1   
21                   EWS Match          object     

In [67]:
df.drop(columns_to_drop, axis=1).to_csv("../data/lovely-rita-clean.csv")